text/1004-transactions-api.rst

```diff
@@ -0,0 +1,619 @@
+
+1. Repeating read-only queries is always safe, but we don't know which
+   ones are readonly. We tackle this below.
+2. Repeating non-readonly queries can be dangerous: they may be applied
+   twice.
+
+There are couple of ideas to differentiate read-only and mutable
+queries:
+
+1. We can expose "read-only" flag in ``Prepare`` (and it's always safe
+   to retry before ``Execute`` happens)
+2. Enable connection mode that forbids mutating queries outside of
+   transaction and retry all (read-only) queries
+3. Do (2) and have some method to override it on per-method-call basis
+4. Or vice versa add ``pool.read_only_query()`` method
+5. Add ``pool.read_only().query()``
+6. Add ``pool.with_options({readonly: true}).query()``
+
+Note: methods (2-6) are also helpful for working with primary/replica
+installations. But probably only last two would allow full power, as they allow
+``pool.read_only(primary=true)`` (i.e. in case you need read-only transaction
+that can't go to a replica).
+
+This issue can be solved by a later RFC.
+
+
+Learning Curve
+==============
+
+This complicates learning curve, but:
+
+1. This is already a problem in Postgres, there a lot of people who
+   ignore the issue for pet projects and a lot of startups having "normal"
+   rate or 500 errors at any point of time, while it's preventable.
+2. Repeating transactions would make less incentive to keep transactions
+   open for a long time (e.g. while accessing slow network resources like
+   external API), which is a problem of itself.
+3. Even if we never have failed concurrent updates we would want
+   seamless reconnect on connection failures (i.e. server restart,
+   primary/replica change, etc.)
+4. To make learning curves shorter I think we should intentionally
+   inject failures. This is needed so that users quickly find out that side
+   effects of their transactions are in effect several times.
+
+So while increasing learning curve, we fix heisenbugs and simplify
+operations.
+
+Failure Injection
+-----------------
+
+The following is proposed to be done by default:
+
+Collect statitics of how many queries are executed in the previous
+second and on each new request trigger a failure with the probability of
+``1/n`` where ``n`` is the number of requests in the previous second. We
+still need to figure out whether ``n`` counts queries, transactions, or
+mutable queries/transactions (and have a list of exceptions, perhaps:
+dump+restore+migrations).
+
+The idea is that there will be ~1 failure per second. So on local
+instance when testing manually it would hit almost every request (which
+is fine as repeating them shouldn't be prohibitively costly). But under
+a huge load of thousands requests per second, one retry per second
+doesn't influence anything so even for production and/or benchmarks this
+is fine.
+
+I think it should be disabled by an explicit command-line argument like
+``--disable-failure-injection`` but might be tweaked with configuration
+settings?
```

Comment:
I like the idea of injecting failures a lot, but just not as something enabled by default. Fundamentally, RDBMS should make all efforts to make operation succeed safely and nicely. Injecting errors would break code that uses the `tryTransaction` API and systems that are just not created for with high-loads in mind.

Instead I suggest adding a session-level setting so that users can enable failure injection in their staging apps and test frameworks.

Reply:
> I like the idea of injecting failures a lot, but just not as something enabled by default. Fundamentally, RDBMS should make all efforts to make operation succeed safely and nicely. Injecting errors would break code that uses the `tryTransaction` API and systems that are just not created for with high-loads in mind.
> 

This isn't for "high-load" this is about concurrency. Technically any amount of concurrency will get "rare" errors. People using `try_transaction` should be prepared for that.

There are two use cases where it's bad:
1. Handling non-concurrent bulk data insert, i.e. initial preload of the data.
2. Users executing things manually: REPL or IPython

> Instead I suggest adding a session-level setting so that users can enable failure injection in their staging apps and test frameworks.

Well, that's possible but my point that this is not enough. If we recommend `retry` everywhere, users will probably use that, but don't realize how repeating side-effects will create a problem.

Anyway, let's create a separate issue/discussion for this. I think it's not a problem to update RFC later, this section is important to keep in RFC, though.